toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go

```diff
@@ -71,133 +152,236 @@ func CustomizeImageWithConfigFile(buildDir string, configFile string, imageFile
-			} else {
-				err = fmt.Errorf("failed to clean-up (%s): %w", rawImageFile, cleanupErr)
+		if inputIsoArtifacts != nil {
+			cleanupErr := inputIsoArtifacts.cleanUp()
+			if cleanupErr != nil {
+				if err != nil {
+					err = fmt.Errorf("%w:\nfailed to clean-up iso builder state:\n%w", err, cleanupErr)
+				} else {
+					err = fmt.Errorf("failed to clean-up iso builder state:\n%w", cleanupErr)
+				}
 			}
 		}
 	}()
 
-	logger.Log.Infof("Creating raw base image: %s", rawImageFile)
-	err = shell.ExecuteLiveWithErr(1, "qemu-img", "convert", "-O", "raw", imageFile, rawImageFile)
+	err = customizeOSContents(imageCustomizerParameters)
+	if err != nil {
+		return fmt.Errorf("failed to customize raw image:\n%w", err)
+	}
+
+	err = convertWriteableFormatToOutputImage(imageCustomizerParameters, inputIsoArtifacts)
 	if err != nil {
-		return fmt.Errorf("failed to convert image file to raw format:\n%w", err)
+		return fmt.Errorf("failed to convert customized raw image to output format:\n%w", err)
+	}
+
+	logger.Log.Infof("Success!")
+
+	return nil
+}
+
+func convertInputImageToWriteableFormat(ic *ImageCustomizerParameters) (*LiveOSIsoBuilder, error) {
+	logger.Log.Infof("Converting input image to a writeable format")
+
+	if ic.inputImageFormat == ".iso" {
+
+		inputIsoArtifacts, err := createIsoBuilderFromIsoImage(ic.buildDir, ic.buildDirAbs, ic.inputImageFile)
+		if err != nil {
+			var cleanUpError error
+			if inputIsoArtifacts != nil {
+				cleanUpError = inputIsoArtifacts.cleanUp()
+			}
+			if cleanUpError == nil {
+				return nil, fmt.Errorf("failed to load input iso artifacts:\n%w", err)
+			} else {
+				return nil, fmt.Errorf("failed to load input iso artifacts:\n%w\nclean-up error:\n%w", err, cleanUpError)
+			}
+		}
+
+		// If the input is a LiveOS iso and there are OS customizations
+		// defined, we create a writeable disk image so that mic can modify
+		// it. If no OS customizations are defined, we can skip this step and
+		// just re-use the existing squashfs.
+		if ic.customizeOSPartitions {
+			err = inputIsoArtifacts.createWriteableImageFromSquashfs(ic.buildDir, ic.rawImageFile)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create writeable image:\n%w", err)
+			}
+		}
+		return inputIsoArtifacts, nil
+
+	} else {
+		logger.Log.Infof("Creating raw base image: %s", ic.rawImageFile)
+		err := shell.ExecuteLiveWithErr(1, "qemu-img", "convert", "-O", "raw", ic.inputImageFile, ic.rawImageFile)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert image file to raw format:\n%w", err)
+		}
+	}
+
+	return nil, nil
+}
+
+func customizeOSContents(ic *ImageCustomizerParameters) error {
+
+	// If the user has defined customizations, then should proceed as usual.
+	// However, there are no customizations, and the input is an iso, we
+	// should just return - but call out any command line switches that might
+	// have requested unsupported changes.
+	if !ic.customizeOSPartitions && ic.inputImageFormat == ".iso" {
```

Comment:
Should we also error out if the input is an ISO and the config has a `.storage` value?